cmd/panfs-multipart.go

```diff
@@ -872,33 +1003,68 @@ func (fs *PANFSObjects) AbortMultipartUpload(ctx context.Context, bucket, object
 		return toObjectErr(err, bucket)
 	}
 
+	handleStatMetaFileError := func(e error) error {
+		// caller is responsible to check that e is not nil
+		if err == errFileNotFound || err == errFileAccessDenied {
+			return InvalidUploadID{Bucket: bucket, Object: object, UploadID: uploadID}
+		}
+		return toObjectErr(err, bucket, object)
+	}
+
+	uploadIDDir := fs.getUploadIDDir(bucketPath, object, uploadID)
+	_, err = fsStatFile(ctx, pathJoin(uploadIDDir, fs.metaJSONFile))
+	if err != nil {
+		return handleStatMetaFileError(err)
+	}
+
 	fs.appendFileMapMu.Lock()
 	// Remove file in tmp folder
 	file := fs.appendFileMap[uploadID]
-	if file != nil {
-		fsRemoveFile(ctx, file.filePath)
+	if file == nil {
+		file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
+		fs.appendFileMap[uploadID] = file
 	}
-	delete(fs.appendFileMap, uploadID)
 	fs.appendFileMapMu.Unlock()
 
-	uploadIDDir := fs.getUploadIDDir(bucketPath, bucket, object, uploadID)
-	// Just check if the uploadID exists to avoid copy if it doesn't.
+	if err = file.flock.Lock(); err != nil {
+		return InvalidUploadID{Bucket: bucket, Object: object, UploadID: uploadID}
+	}
+	defer func() {
+		if err == nil {
+			file.flock.Unlock()
+			// remove lock file
+			fsRemoveFile(ctx, fs.getMultipartLockFile(bucketPath, object, uploadID))
+			// Delete append file from memory only if no error occurred during mkdir and rename on next steps
+			fs.appendFileMapMu.Lock()
+			delete(fs.appendFileMap, uploadID)
+			fs.appendFileMapMu.Unlock()
+		} else {
+			file.flock.Unlock()
+		}
+	}()
+
+	// ensure that uploadId dir is still here
 	_, err = fsStatFile(ctx, pathJoin(uploadIDDir, fs.metaJSONFile))
 	if err != nil {
-		if err == errFileNotFound || err == errFileAccessDenied {
-			return InvalidUploadID{Bucket: bucket, Object: object, UploadID: uploadID}
-		}
-		return toObjectErr(err, bucket, object)
+		return handleStatMetaFileError(err)
 	}
+	fsRemoveFile(ctx, file.filePath)
 
 	// Purge multipart folders
-	fsTmpObjPath := pathJoin(bucketPath, panfsS3TmpDir, fs.nodeDataSerial, mustGetUUID())
-	defer fsRemoveAll(ctx, fsTmpObjPath) // remove multipart temporary files in background.
+	tempDir := pathJoin(fs.getTempDir(bucketPath), mustGetUUID())
+	if err = MkdirAll(tempDir, 0o755); err != nil {
+		return toObjectErr(err, bucket, object)
+	}
+	fsTmpObjPath := pathJoin(tempDir, mustGetUUID())
+	defer fsRemoveAll(ctx, fsTmpObjPath)
 
-	Rename(uploadIDDir, fsTmpObjPath)
+	err = Rename(uploadIDDir, fsTmpObjPath)
+	if err != nil {
+		return toObjectErr(err, bucket, object)
+	}
 
 	// It is safe to ignore any directory not empty error (in case there were multiple uploadIDs on the same object)
-	fsRemoveDir(ctx, fs.getMultipartSHADir(bucketPath, bucket, object))
+	fsRemoveDir(ctx, fs.getMultipartSHADir(bucketPath, object))
```

Comment:
If there's another multipart upload for this object running concurrently, we are breaking it by removing that directory.

Reply:
Theoretically this is possible if abort removes that directory in a short window when other goroutine creates an uploadId directory but does not creates a `panfs.json` file inside.
As we have periodic cleanup procedure - we can do all cleaning stuff in that procedure then